src/control/events/pubsub_test.go

```diff
@@ -234,32 +235,64 @@ func TestEvents_PubSub_Debounce_NoCooldown(t *testing.T) {
 }
 
 func TestEvents_PubSub_Debounce_Cooldown(t *testing.T) {
-	log, buf := logging.NewTestLogger(t.Name())
-	defer test.ShowBufferOnFailure(t, buf)
+	test := func(t *testing.T, iter int) bool {
+		log, buf := logging.NewTestLogger(t.Name())
+		defer test.ShowBufferOnFailure(t, buf)
+
+		ps := NewPubSub(test.Context(t), log)
+		defer ps.Close()
+
+		debounceType := RASSwimRankDead
+		debounceCooldown := 25 * time.Millisecond
+		tally := newTally(3)
+
+		ps.Subscribe(RASTypeStateChange, tally)
+		ps.Debounce(debounceType, debounceCooldown, func(ev *RASEvent) string {
+			// Debounce should match all events sent in test.
+			return fmt.Sprintf("%d", ev.Rank)
+		})
+
+		// Send 16 (var j) events per incarnation (var i). We should only see this event
+		// three times as a new event will only be registered after the cooldown timer
+		// expires at the end of each incarnation loop iteration.
+		for i := 0; i < 3; i++ {
+			log.Debugf("start loop %d", i+1)
+
+			e := mockSwimRankDeadEvt(1, uint32(i+1))
+			t := time.Now()
+			for j := 0; j < 16; j++ {
+				l := time.Since(t)
+				if l >= debounceCooldown {
+					// Test loop was stalled, print warning
+					log.Noticef("test loop stalled for %s, restart iteration", l)
+					return true
```

Comment:
Not sure to understand why you return true.
Maybe continue could be more suited ?

Reply:
The "test" function returns true to indicate a restart which is detected on L294 and the test counter is decremented so the whole test runs again (resulting in teardown and setup). When a false expiry is triggered by the VM lockup all the test results are invalid so it should restart from the beginning. This is an acceptable solution because the false expiry seems to occur only once in every thousand test iterations or less.

Reply:
Thanks for the clarification :)